src/awkward/_connect/pyarrow.py

```diff
@@ -196,38 +196,6 @@ def num_fields(self):
         pyarrow.duration("ns"): (False, np.dtype("m8[ns]")),
     }
 
-if not ak._util.numpy_at_least("1.17.0"):
-
-    def packbits(bytearray, lsb_order=True):
-        if lsb_order:
-            if len(bytearray) % 8 == 0:
-                ready_to_pack = bytearray
-            else:
-                ready_to_pack = numpy.empty(
-                    int(numpy.ceil(len(bytearray) / 8.0)) * 8,
-                    dtype=bytearray.dtype,
-                )
-                ready_to_pack[: len(bytearray)] = bytearray
-                ready_to_pack[len(bytearray) :] = 0
-            return numpy.packbits(ready_to_pack.reshape(-1, 8)[:, ::-1].reshape(-1))
-        else:
-            return numpy.packbits(bytearray)
-
-    def unpackbits(bitarray, lsb_order=True):
-        ready_to_bitswap = numpy.unpackbits(bitarray)
-        if lsb_order:
-            return ready_to_bitswap.reshape(-1, 8)[:, ::-1].reshape(-1)
-        else:
-            return ready_to_bitswap
-
-else:
-
-    def packbits(bytearray, lsb_order=True):
-        return numpy.packbits(bytearray, bitorder=("little" if lsb_order else "big"))
-
-    def unpackbits(bitarray, lsb_order=True):
-        return numpy.unpackbits(bitarray, bitorder=("little" if lsb_order else "big"))
```

Comment:
I couldn't find the cause of the `numpy_at_least` here using `git blame`. @jpivarski is there a reason for this switch that I need to consider? Or can we just use NumPy's built-in version

Reply:
As an alternative, you could put the cute slice work-around in an nplike implementation of `packbits`/`unpackbits`. Then we can use the future-proof interface everywhere (a `bitorder` argument, rather than `lsb_order`, which is designed around the `BitMaskedArray` class) and remove it from one spot—the Numpy nplike—when we increase our minimum NumPy.

To summarize, these are the important version thresholds:

  * 1.13.1: NEP 13 introduced
  * 1.17.0: NEP 18 became a default
  * 1.17.0: `packbits`/`unpackbits` got the `bitorder` argument

Hmmm. Maybe 1.17.0 could be a good threshold. It was July 26, 2019. (1.13.1 was July 6, 2017, so we'd be moving up 2 years. Awkward 0 started in June 2018, requiring NumPy 1.13.1 for ufuncs to work, so moving up to 1.17.0 now would still be a bigger gap now than it was at the start.)

In fact, NumPy 1.13.1 isn't even listed on https://endoflife.date/numpy or https://numpy.org/neps/nep-0029-deprecation_policy.html. By setting the minimum to NumPy 1.17.0, we're still picking a minimum version that isn't supported by the NumPy team itself (i.e. we accept all versions that NumPy supports, and then some), but at least it's on the table with a date next to it. Also, NumPy 1.13.1 technically supports Python 3.6, which we don't anymore.

New plan: we set the minimum NumPy to 1.17.0. Does that sound good to you? I can also put out a question on Slack #awkward-uproot.

Reply:
Ah, but doing something like that, we should do it on a minor version number, so that these things are easier to track by hand (if need be; I know, pip would do this one automatically because it's a strict dependency).

So we should put "minimum NumPy version >= 1.17.0" on the table for Awkward 2.1.0 and for now have `packbits`/`unpackbits` go through nplike with the backward-compatible workaround in one place, slated for removal in 2.1.0.

Reply:
> Yes, it's because the bitorder argument was only introduced in NumPy 1.17.0

Huh that didn't render on mobile when I was looking, but that was my assumption... wierd!

> Also, NumPy 1.13.1 technically supports Python 3.6, which we don't anymore.

Yep, I am enthusiastic in my support for dropping 1.13 in favour of 1.17.

> packbits/unpackbits go through nplike with the backward-compatible workaround in one place,

Yep, this is an easy fix for now.